pkg/proc/gdbserial/undo.go

```diff
@@ -158,3 +160,107 @@ func UndoIsRecording(recordingFile string) (result bool, err error) {
 
 	return bytes.Equal(marker, data), nil
 }
+
+// Fetch the output of a udbserver get_info command, split on ; and , characters.
+//
+// This is not (currently) implementing a proper parse of the data returned, just making it more
+// convenient to search.
+func undoGetInfo(conn gdbConn) []string {
+	info, err := conn.undoCmd("get_info")
+	if err != nil {
+		panic("Failed to retrieve Undo stop info.")
+	}
+	splitter := func(c rune) bool {
+		return c == ';' || c == ','
+	}
+	return strings.FieldsFunc(info, splitter)
+}
+
+func undoGetLogExtent(p *gdbProcess) (uint64, uint64) {
+	extent, err := p.conn.undoCmd("get_log_extent")
+	if err != nil {
+		panic("Failed to retrieve Undo log extent.")
+	}
+	bbcounts := strings.Split(extent, ",")
+	bbcount_min, err := strconv.ParseUint(bbcounts[0], 16, 64)
+	if err != nil {
+		panic("Failed to parse Undo log minimum.")
+	}
+	bbcount_max, err := strconv.ParseUint(bbcounts[1], 16, 64)
+	if err != nil {
+		panic("Failed to parse Undo log maximum.")
+	}
+	return bbcount_min, bbcount_max
+}
+
+// Fetch whether the replay session is currently at the end of recorded history.
+func UndoAtEndOfHistory(p *gdbProcess) bool {
+	info_fields := undoGetInfo(p.conn)
+	for _, value := range info_fields {
+		if value == "has_exited" || value == "at_event_log_end" {
+			return true
+		}
+	}
+	return false
+}
+
+// Fetch the process exit code (or zero, if not applicable) from the recording.
+func UndoGetExitCode(p *gdbProcess) int {
+	exit_code := int8(0)
+	info_fields := undoGetInfo(p.conn)
+
+	for idx, value := range info_fields {
+		if value != "has_exited" {
+			continue
+		}
+
+		// Exit status, encoded as hex, follows the has_exited string.
+		exit_status, err := strconv.ParseInt(info_fields[idx+1], 16, 16)
+		if err != nil {
+			panic("Invalid exit code.")
+		}
+
+		// Convert exit status into the form Delve usually reports - positive integer for a
+		// normal exit, negative signal number if terminated by a signal.
+		if exit_status&0x7f != 0 {
+			exit_code = -int8(exit_status & 0x7f)
+		} else {
+			exit_code = int8(exit_status >> 8)
+		}
+		break
+	}
+
+	return int(exit_code)
+}
+
+func undoWhen(p *gdbProcess) (string, error) {
+	resp, err := p.conn.undoCmd("get_time")
+	if err != nil {
+		return "", err
+	}
+
+	// We have received a comma-separated list of hex numbers.
+	time_parts := strings.Split(resp, ",")
+
+	// First component is bbcount.
+	bbcount, err := strconv.ParseUint(time_parts[0], 16, 64)
+	if err != nil {
+		return "", err
+	}
+
+	// Second component is PC.
+	pc, err := strconv.ParseUint(time_parts[1], 16, 64)
+	if err != nil {
+		return "", err
+	}
+
+	// Calculate our percentage through available history.
+	bbcount_min, bbcount_max := undoGetLogExtent(p)
+	progress := uint64(0)
+	if bbcount_min != bbcount_max {
+		progress = ((bbcount - bbcount_min) * 100) / (bbcount_max - bbcount_min)
+	}
+
+	result := fmt.Sprintf("[replaying %d%% %d:0x%x]", progress, bbcount, pc)
```

Comment:
UDB uses `start` and `end` when the current bbcount is at the start or end of history respectively. If you prefer to use a numeric percentage consistently, then I think 100% would be more intuitive for the case where `bbcount_min == bbcount_max`.

Reply:
I wanted to show `start` and `end` but Delve doesn't invoke the relevant status printing at the extremes of history, so nothing we do here can make a difference.

I can set the variable to 100 if you think it'll be better future-proofing but I'm afraid it's still not going to get displayed until some kind of more invasive changes are made (i.e. I suspect we'd need to change code outside the usual backend area).

Reply:
(hmmm, I'll update this slightly - at the start of time for a non-zero-length recording we would get an opportunity to display `start`)

Reply:
I fiddled with this a little and then decided to leave it alone for now.

You'll see 0% at the start of a recording with non-zero length.  For a recording of zero length I think you just won't see this at all - but I'd hope to revisit it later when we've had some more customer exposure.

Reply:
I've left this as-is, except for updating error handling (previously `undoGetLogExtent()` wasn't returning errors to this code) and making the progress default to 100% if the bbcounts are equal (though sadly that won't be visible until this gets revisited - Delve really just wants to print exit messages at the end of history, even when it doesn't make sense for us)

Reply:
Fine to leave this as-is.